livekit-agents/livekit/agents/pipeline/pipeline_agent.py

```diff
@@ -658,84 +737,27 @@ def _commit_user_question_if_needed() -> None:
-        if is_using_tools and not interrupted:
-            assert isinstance(speech_handle.source, LLMStream)
-            assert (
-                not user_question or speech_handle.user_commited
-            ), "user speech should have been committed before using tools"
-
-            # execute functions
-            call_ctx = AgentCallContext(self, speech_handle.source)
-            tk = _CallContextVar.set(call_ctx)
-            self.emit("function_calls_collected", speech_handle.source.function_calls)
-            called_fncs_info = speech_handle.source.function_calls
-
-            called_fncs = []
-            for fnc in called_fncs_info:
-                called_fnc = fnc.execute()
-                called_fncs.append(called_fnc)
-                logger.debug(
-                    "executing ai function",
-                    extra={
-                        "function": fnc.function_info.name,
-                        "speech_id": speech_handle.id,
-                    },
-                )
-                try:
-                    await called_fnc.task
-                except Exception:
-                    pass
-
-            self.emit("function_calls_finished", called_fncs)
-            _CallContextVar.reset(tk)
-
-            tool_calls = []
-            tool_calls_results_msg = []
-
-            for called_fnc in called_fncs:
-                # ignore the function calls that returns None
-                if called_fnc.result is None:
-                    continue
-
-                tool_calls.append(called_fnc.call_info)
-                tool_calls_results_msg.append(
-                    ChatMessage.create_tool_from_called_function(called_fnc)
-                )
+    async def _play_speech(self, speech_handle: SpeechHandle) -> None:
+        try:
+            await speech_handle.wait_for_initialization()
+        except asyncio.CancelledError:
+            return
 
-            if tool_calls:
-                extra_tools_messages.append(ChatMessage.create_tool_calls(tool_calls))
-                extra_tools_messages.extend(tool_calls_results_msg)
+        await self._agent_publication.wait_for_subscription()
 
-                chat_ctx = speech_handle.source.chat_ctx.copy()
-                chat_ctx.messages.extend(extra_tools_messages)
+        if speech_handle.synthesis_handle.interrupted:
+            return
 
-                answer_llm_stream = self._llm.chat(
-                    chat_ctx=chat_ctx,
-                    fnc_ctx=self._fnc_ctx,
-                )
-                answer_synthesis = self._synthesize_agent_speech(
-                    speech_handle.id, answer_llm_stream
-                )
-                # replace the synthesis handle with the new one to allow interruption
-                speech_handle.synthesis_handle = answer_synthesis
-                play_handle = answer_synthesis.play()
-                await play_handle.join()
+        await self._wait_and_commit_user_question(speech_handle)
```

Comment:
The diff looks weird - here I just moved that whole wait for user message to its own awaitable method.